contrib/cloud.google.com/go/pubsub.v1/pubsub.go

```diff
@@ -0,0 +1,59 @@
+package pubsub
+
+import (
+	"context"
+
+	"cloud.google.com/go/pubsub"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// Publish publishes a message on the specified topic, waits for publishing to complete, and returns the result.
+// It is functionally equivalent to (*pubsub.Topic).Publish(ctx, msg).Get(ctx), but it also ensures that the datadog
+// tracing metadata is propagated as attributes attached to the published message.
+func Publish(ctx context.Context, t *pubsub.Topic, msg *pubsub.Message) (serverID string, err error) {
+	span, ctx := tracer.StartSpanFromContext(
+		ctx,
+		"pubsub.publish",
+		tracer.ResourceName(t.String()),
+		tracer.SpanType(ext.SpanTypeMessageProducer),
+		tracer.Tag("message_size", len(msg.Data)),
+		tracer.Tag("ordering_key", msg.OrderingKey),
+	)
+	defer func() {
+		span.Finish(tracer.WithError(err))
+	}()
+	if msg.Attributes == nil {
+		msg.Attributes = make(map[string]string)
```

Comment:
Can you please let me know why this was needed? Making a new map every time can cost in hotpaths. Not sure if it applies here but still... Why is `nil` bad?

I mean, when this is `nil`, we could avoid setting the tag, and also injecting into the carrier.

Reply:
> Why is nil bad?

Because TextMapCarrier will [panic](https://github.com/DataDog/dd-trace-go/blob/6d7c52cd8ab1ea309fd54c1a2ef472fadb84429c/ddtrace/tracer/textmap.go#L51-L53) with a nil map.

Reply:
> I mean, when this is nil, we could avoid setting the tag, and also injecting into the carrier.

Not sure that makes sense. If the user does not want tracing, they should not add the middleware.
If they want tracing, the middleware needs to allocate the map to propagate the headers, there's no other way.
FWIW this is basically the same thing that happens with http requests, so I think we shouldn't be very concerned with this.

Reply:
Makes complete sense. Let's leave it.